### DISCUSSION TITLE
BLTouch 3.0 and 3.1 Info/ Possible fix for M112 failed probing

### DISCUSSION THREAD

**Discussion by DeathDonky** [Jyers/Marlin#603] [General Discussion]
Really new to this but I did come across a "fix" that may be worth mentioning... I am using an Ender 3v2 with this flipping AMAZING firmware. Thank you 1: for including your SOURCE codes because that allowed me, a nobody in this world, to learn. I really appreciate that. It allowed me to not only learn but 2: customize the ender with your great features to my parts. Love it. Please keep going. anyway...

With the ender 3v2 I'm using a BLTouch v3.1. it is a genuine BLTouch. Upon installing many other firmware and finally settling with this one I came across an issue where about only 30% of my pre-print probs would complete. for example, I used a 3x3, 5x5, and currently a 7x7(custom) probe grid before each print. (I'm an accuracy nut...) In either grid setting... I might get through 3 points, might get through 8, might get to the very last one but only about 30% of the time would it complete. This happened on stock to custom firmware and I kept chasing the issue. When it would fail id get the "M112 Kill command restart printer required"  Now Jyers has that help page which shows how to setup octoprint to ignore these failed probes and just carry on with the print anyway. It would get the error and start blinking like normal and then just move over to its warmup position and print. My issue with this is I didn't have the peace of mind that my mesh was proper. yes the print would carry on and I no longer had to power cycle but that wasn't the fix I needed. 

Come to find out in the config adv file there is a setting for "#define BLTOUCH_SET_5V_MODE". ONLY IF ITS A BLTOUCH V3.0 OR 3.1 DOES THIS PERTAIN TO. 

I am happy to report i now have a 100% probe success rate now and have probed the bed every bit of 200+ times. Remember before id get maybe a 30% success. Now all my probes are passing and i have octoprint set back up to not ignore errors. 

Idk if this is useful... I hope it is as i know i want the only one chasing this... 

On a side note i have tested both with and without the HS probing enabled. Both cases 100% success rate.

**Comment by MrDragon420** [comment]
I also have probeing issues... will have to try this

**Reply by gothcha** [reply]
If you are using the 32bit boards 4.2.2 & 4.2.7 (std on V2) the ST micro is NOT 5V tolerant. You are likely to fry the processor if you enable 5V mode!!! There is a thread discussing this and possible safe fixes by using the ZMin port or adding a pull-up resistor

**Reply by gothcha** [reply]
See #192

**Reply by DeathDonky** [reply]
In the configuration.adv a line reads 
“  * V3.0 or 3.1: Set default mode to 5V mode at Marlin startup. “

I uncommented the enable 5v right under that and all seems fine. Been printing 3 days solid with it under them settings.

**Reply by gothcha** [reply]
OK I am just letting you know that it is not recommended and is likely to reduce the life of your board. Yes It will work - until it doesn’t! There are plenty of links detailing this issue if you Google it.

**Reply by DeathDonky** [reply]
What does it do to the processor? I spent a couple days trying to look into the issue on google and came across very seldom results that seemed to pertain to my problem exactly on this printer. I’m hard pressed to modify my board by adding a resistor and hard pressed to “ignore” potential issues using that octoprint work around as it didn’t aid in the success rate of probing it simply ignored it when it happened. 

I asked for help originally here (https://www.reddit.com/r/ender3v2/comments/mw6k6u/octoprintbltouch_probe_failingm112_restart/?utm_source=share&utm_medium=ios_app&utm_name=iossmf) 

And with what was mentioned through where I was directed here I came across another bunch you seemed to have success with this fix. Again only with v3.0 and 3.1 bltouch. 

Maybe also worth mentioning my plugs for the bltouch aren’t split anywhere. I don’t have a 3+2 that goes to the board it’s a dedicated plug. 

You seem to have mistaken what I said above as an “I’m right your wrong” situation... I’m sorry if I came across that way but I’m new to this hobby... I’m just trying to get all the information and ensure it’s the RIGHT info is all man... I’m not trying to sound conceded or like I know it all I just want to learn the right stuff. Sorry if I came off that way...

**Reply by gothcha** [reply]
Hi no problem I am trying to save you damage. Check the antclabs data sheet, you will see that they say not to enable the 5V mode on 3.3V systems, which the 4.2.2 is.
<img width="578" alt="Screenshot 2021-04-26 at 14 53 39" src="https://user-images.githubusercontent.com/80574048/116094293-53bf7800-a69f-11eb-86f2-4c1845914944.png">

https://www.antclabs.com/bltouch-v3

The port pin on the processor is likely to be damaged by over voltage. So I cannot advise that you use this method, however it is your decision.

**Reply by adam3654** [reply]
I saw this thread and got excited because I have these intermittent issues as well. 

I jumped the gun and uncommented the 5v line last night, and the board is alive, but seeing the negative response, I reverted and recompiled to go back to the original setup. 

I verified that according to the datasheet for the stm32 chip on the 4.2.x boards, the BLTouch in pin (PB1) is **NOT** 5v tolerant. There are some pins on the chip that are if the internal pull-up/down is disabled.

**Reply by DeathDonky** [reply]
![15C2788E-5D3B-404E-B3CA-3ACA7E48BDA7](https://user-images.githubusercontent.com/83191001/116117417-d33e4e80-a681-11eb-98f0-a6d0a773faae.png)


Again PLEASE don’t take this the wrong way but if what you guys are saying is true can you please explain what the info in the screenshot means and when it should be used?

**Reply by adam3654** [reply]
Refer to the line in the screenshot that begins with the word "Danger"

> ![15C2788E-5D3B-404E-B3CA-3ACA7E48BDA7](https://user-images.githubusercontent.com/83191001/116117417-d33e4e80-a681-11eb-98f0-a6d0a773faae.png)
> 
> 
> Again PLEASE don’t take this the wrong way but if what you guys are saying is true can you please explain what the info in the screenshot means and when it should be used?

**Reply by DeathDonky** [reply]
Right. And the “controller” would be the processor? What confuses me is the line right below. I’m using a 3.1 and thus uncommented

**Reply by adam3654** [reply]
I agree that the wording is weird, but according to the datasheet for the chip, 5v->PB1 = No Bueno

**Reply by DeathDonky** [reply]
https://www.st.com/resource/en/datasheet/stm32f103c8.pdf#page37

Is this the RIGHT data sheet?? I just google the “stm32f103 data sheet” and got that?

**Reply by DeathDonky** [reply]
Stm32f103 is on what I’m assuming is the processor of my board

**Reply by gothcha** [reply]
The 5V mode was included by BLTouch for the old 8bit Atmel boards, these are powered from 5V, so are safe. This is why the option is available. But it is not for the new 32bit 3.3V boards.
This is all I am going to say on the matter. Good luck!

**Reply by adam3654** [reply]
https://datasheet.octopart.com/STM32F103RET6-STMicroelectronics-datasheet-62336676.pdf

This is the datasheet for the chip

**Reply by DeathDonky** [reply]
So let me clarify because again new here. What your saying is before things were on a 3.3 volt system but now these new processors are running on 5v or can at least take that voltage?(the processor named above with the respective data sheet) Thus allowing that voltage to be “safe” in theory? 

let me specify no one but myself is responsible for “damages” to my printer but me. I just find it really hard to believe though a modification such as soldering is required on a product listed as plug and play minus a firmware flash. It even has its own port on the board labeled BLT I believe it was. I made sure orientation was correct.

**Reply by adam3654** [reply]
The old 8 bit boards were 5 volt, the new stm32 based boards (4.2.x) are 3.3 volt. They do have some input pins that can be set up to handle a 5 volt input signal, but the BLTouch pin isn't one of them.

**Reply by DeathDonky** [reply]
So this explains why my cr10v3 never had this issue... it’s using an identical v3.1 bltouch and never had these probing issues. 

It too has the 5v line uncommented. 

So I’m at work and the printer in question that I am currently over volting is at home... printing away. Has been for the last 3 days now. I really hope I’m not harming it right now but worst case scenario I could replace the board if needed right? 

I’m pretty embarrassed and not gonna lie feel kinda retarded rn LOL thanks for you guys time and input/dealing with me

**Reply by adam3654** [reply]
It should be ok, or at least no worse since it homed z before the start of the print. I'm pretty sure it just effects the input pin. 

This discussion forced me to dig a little deeper into the source than I have before. There really is no dumb questions, I'm going to have a ton of questions as I learn. I'm just getting started with 3d printing.

**Reply by DeathDonky** [reply]
Here’s a question... if the processor is powered by 3.3v and THATS what it’s sending to the bltouch... would the return signal still be 5v? Does uncommenting that line ACTUALLY increase voltages or “allow” signals to be of that magnitude to still be “ok”? If any of that makes sense.

**Reply by DeathDonky** [reply]
Not only did this “fix” my success rate of probing the “issue with firmware” pop up quit showing its face on octoprint even prior to testing the success rates.

**Reply by gothcha** [reply]
No there is a separate 5V supply sent to the BLTouch. Which is why if this is sent to the processor it will damage it.  Just take our word for it, we do know what we are talking about.

**Reply by DeathDonky** [reply]
Hey gents. So i came home and after my print finished I did some more probing with a multimeter on the bltouch. It does not seem to matter if the "enable 5v" line is commented or uncommented it still gets 5v either way. This didn't change any of the voltages going to or from the BLTouch.

**Reply by DeathDonky** [reply]
in both cases any time the BLTouch was triggered it sent less than a volt. Again both with and without that line commented. 
on mine that was sent through the yellow(middle most wire) on the BLTouch itself.

**Reply by t3chguy** [reply]
@DeathDonky chances are the capacitance on your multimeter/probes is high enough that due to how short the signal edge is the multimeter will under-measure, these sorts of signals need to be measured on an oscilloscope.

**Reply by DeathDonky** [reply]
In either case the voltage received to the bltouch was a constant 5v... across black and white wire.  it didn’t change regardless of the configuration.adv setting and the multimeter was capable of that at least because it was constant.

**Reply by t3chguy** [reply]
Yes, that voltage comes from the 5V regulator.
The Marlin config is how to configure the operation of the BLTouch in software, this changes what the BLTouch does with its output pin in its own hardware.
in 5V mode it changes how the output pin works. in 5V mode it sends 5V/GND signals, without that mode it is an open drain output so it can be used with a lower voltage system like 3v3 of the ARM chip as the ARM chip pulls the GPIO up to 3.3v via an internal resistor and the BLTouch grounds it or leaves it floating depending on the state. No 5V -> ARM which can (and will eventually) cause damage. The current is low enough that the damage will not be immediate or quick but nearly certain.

**Reply by gothcha** [reply]
@t3chguy Thank you for the support :-)

**Reply by DeathDonky** [reply]
That really sucks. Looks like octoprint will be going away from the ender for the time being then..
